tests/sample_size/test_metrics.py

```diff
@@ -76,6 +77,74 @@ def test_boolean_metric_get_probability_too_small(self):
         )
 
+    @patch("scipy.stats.uniform.rvs")
+    @patch("scipy.stats.nct.rvs")
+    def test_boolean_metric_generate_p_value_two_variants_true_null(self, mock_nct, mock_uniform):
+        true_alt = False
+        size = 200
+        variants = 2
+        pval = 1
+        mock_uniform.return_value = pval
+
+        boolean = BooleanMetric(self.DEFAULT_PROBABILITY, self.DEFAULT_MDE)
+        p = boolean.generate_p_value(true_alt, size, variants)
+        mock_uniform.assert_called_once_with(0, 1, size=variants - 1)
+        mock_nct.assert_not_called()
+        self.assertEqual(p, pval)
+
+    @patch("sample_size.metrics.BooleanMetric.variance")
+    @patch("scipy.stats.uniform.rvs")
+    @patch("scipy.stats.nct.rvs")
+    def test_boolean_metric_generate_p_value_two_variants_true_alt(self, mock_nct, mock_uniform, mock_variance):
+        true_alt = True
+        size = 200
+        variants = 2
+        z = 0
+        pval = 1
+        nc = np.sqrt(size / 2) * self.DEFAULT_MDE / self.DEFAULT_MOCK_VARIANCE
+        mock_variance.__get__ = MagicMock(return_value=self.DEFAULT_MOCK_VARIANCE)
+        mock_nct.return_value = z
+
+        boolean = BooleanMetric(self.DEFAULT_PROBABILITY, self.DEFAULT_MDE)
+        p = boolean.generate_p_value(true_alt, size, variants)
+        mock_uniform.assert_not_called()
+        mock_nct.assert_called_once_with(nc=nc, df=2 * (size - 1), size=variants - 1)
+        self.assertEqual(p, pval)
+
+    @patch("scipy.stats.uniform.rvs")
+    @patch("scipy.stats.nct.rvs")
+    def test_boolean_metric_generate_p_value_multiple_variants_true_null(self, mock_nct, mock_uniform):
+        true_alt = False
+        size = 200
+        variants = 5
+
+        boolean = BooleanMetric(self.DEFAULT_PROBABILITY, self.DEFAULT_MDE)
+        pval = np.ones(variants - 1)
+        mock_uniform.return_value = pval
+        p = boolean.generate_p_value(true_alt, size, variants)
+        mock_uniform.assert_called_once_with(0, 1, size=variants - 1)
+        mock_nct.assert_not_called()
+        self.assertTrue((p == pval).all())
+
+    @patch("sample_size.metrics.BooleanMetric.variance")
+    @patch("scipy.stats.uniform.rvs")
+    @patch("scipy.stats.nct.rvs")
+    def test_boolean_metric_generate_p_value_multiple_variants_true_alt(self, mock_nct, mock_uniform, mock_variance):
+        true_alt = True
+        size = 200
+        variants = 5
+        z = np.zeros(variants - 1)
+        pval = np.ones(variants - 1)
+        nc = np.sqrt(size / 2) * self.DEFAULT_MDE / self.DEFAULT_MOCK_VARIANCE
+        mock_variance.__get__ = MagicMock(return_value=self.DEFAULT_MOCK_VARIANCE)
+        mock_nct.return_value = z
+
+        boolean = BooleanMetric(self.DEFAULT_PROBABILITY, self.DEFAULT_MDE)
+        p = boolean.generate_p_value(true_alt, size, variants)
+        mock_uniform.assert_not_called()
+        mock_nct.assert_called_once_with(nc=nc, df=2 * (size - 1), size=variants - 1)
+        self.assertTrue((p == pval).all())
```

Comment:
```suggestion
        assert_array_equal(p, pval)
```
this is more readable and can pass mypy https://numpy.org/doc/stable/reference/generated/numpy.testing.assert_array_equal.html

Reply:
I got the same error even after replacing the type by  `Union[np.typing.NDArray[np.float64], float]`. The only way to have the tests pass is making the returned variable to be an ndarray:
```
  def generate_p_value(self, true_null: bool, sample_size: int, variants: int) -> np.typing.NDArray[np.float64]:
      if true_null:
          return np.array(stats.uniform.rvs(0, 1, size=variants - 1))

      else:
          effect_size = self.mde / float(np.sqrt(2 * self.variance / sample_size))
          z_alt = stats.norm.rvs(loc=effect_size, size=variants - 1)
          return np.array(2 * stats.norm.sf(np.abs(z_alt)))
```
btw I also made changes to the variable names and added a docstring

Reply:
yeah my suggestion includes making the type explicitly in this comment https://github.com/godaddy/sample-size/pull/26#discussion_r875242493, btw can you use `assert_array_equal(p, pval)` instead of `(p == pval).all()`?
with your above code, idk if you need to change anything else if both conditions return np.array. `Union[np.typing.NDArray[np.float64], float]` this means it returns numpy array or a float. But in general i guess it makes more sense to return the same type probably (?) curious to know what others think

BTW it seems like you didn't commit the change yet

Reply:
There's still issue with typing. I just committed the change along with other unit test changes. It failed the checks since I kept getting the error `AttributeError: module 'numpy' has no attribute 'typing'`

Reply:
Numpy has some some [a special type plugin](https://numpy.org/devdocs/reference/typing.html). Perhaps that's the issue?

Reply:
you can check the error under Actions tab https://github.com/godaddy/sample-size/runs/6498722018?check_suite_focus=true
it is the package problem, i think you need to also include lock file?